controllers/bpfman-agent/xdp-ns-program.go

```diff
@@ -0,0 +1,294 @@
+//+kubebuilder:rbac:groups=bpfman.io,namespace=bpfman,resources=xdpnsprograms,verbs=get;list;watch
+
+// BpfProgramReconciler reconciles a BpfNsProgram object
+type XdpNsProgramReconciler struct {
+	NamespaceProgramReconciler
+	currentXdpNsProgram *bpfmaniov1alpha1.XdpNsProgram
+	interfaces          []string
+	ourNode             *v1.Node
+}
+
+func (r *XdpNsProgramReconciler) getFinalizer() string {
+	return r.finalizer
+}
+
+func (r *XdpNsProgramReconciler) getOwner() metav1.Object {
+	if r.appOwner == nil {
+		return r.currentXdpNsProgram
+	} else {
+		return r.appOwner
+	}
+}
+
+func (r *XdpNsProgramReconciler) getRecType() string {
+	return r.recType
+}
+
+func (r *XdpNsProgramReconciler) getProgType() internal.ProgramType {
+	return internal.Xdp
+}
+
+func (r *XdpNsProgramReconciler) getName() string {
+	return r.currentXdpNsProgram.Name
+}
+
+func (r *XdpNsProgramReconciler) getNamespace() string {
+	return r.currentXdpNsProgram.Namespace
+}
+
+func (r *XdpNsProgramReconciler) getNoContAnnotationIndex() string {
+	return internal.XdpNsNoContainersOnNode
+}
+
+func (r *XdpNsProgramReconciler) getNode() *v1.Node {
+	return r.ourNode
+}
+
+func (r *XdpNsProgramReconciler) getBpfProgramCommon() *bpfmaniov1alpha1.BpfProgramCommon {
+	return &r.currentXdpNsProgram.Spec.BpfProgramCommon
+}
+
+func (r *XdpNsProgramReconciler) getNodeSelector() *metav1.LabelSelector {
+	return &r.currentXdpNsProgram.Spec.NodeSelector
+}
+
+func (r *XdpNsProgramReconciler) getBpfGlobalData() map[string][]byte {
+	return r.currentXdpNsProgram.Spec.GlobalData
+}
+
+func (r *XdpNsProgramReconciler) getAppProgramId() string {
+	return appProgramId(r.currentXdpNsProgram.GetLabels())
+}
+
+func (r *XdpNsProgramReconciler) setCurrentProgram(program client.Object) error {
+	var err error
+	var ok bool
+
+	r.currentXdpNsProgram, ok = program.(*bpfmaniov1alpha1.XdpNsProgram)
+	if !ok {
+		return fmt.Errorf("failed to cast program to XdpNsProgram")
+	}
+
+	r.interfaces, err = getInterfaces(&r.currentXdpNsProgram.Spec.InterfaceSelector, r.ourNode)
+	if err != nil {
+		return fmt.Errorf("failed to get interfaces for XdpNsProgram: %v", err)
+	}
+
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+// The Bpfman-Agent should reconcile whenever a XdpNsProgram is updated,
+// load the program to the node via bpfman, and then create a bpfProgram object
+// to reflect per node state information.
+func (r *XdpNsProgramReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&bpfmaniov1alpha1.XdpNsProgram{},
+			builder.WithPredicates(predicate.And(
+				predicate.GenerationChangedPredicate{},
+				predicate.ResourceVersionChangedPredicate{}),
+			),
+		).
+		Owns(&bpfmaniov1alpha1.BpfNsProgram{},
+			builder.WithPredicates(predicate.And(
+				internal.BpfNsProgramTypePredicate(internal.Xdp.String()),
+				internal.BpfProgramNodePredicate(r.NodeName)),
+			),
+		).
```

Comment:
```suggestion
		).
		// Watch for changes in Pod resources in case we are using a container selector.
		Watches(
			&v1.Pod{},
			&handler.EnqueueRequestForObject{},
			builder.WithPredicates(podOnNodePredicate(r.NodeName)),
		).
```

Reply:
Done